engines/e_afalg.c

```diff
@@ -125,7 +125,17 @@ static ossl_inline int io_getevents(aio_context_t ctx, long min, long max,
                                struct io_event *events,
                                struct timespec *timeout)
 {
+#if defined(__NR_io_getevents)
     return syscall(__NR_io_getevents, ctx, min, max, events, timeout);
+#elif __TIMESIZE == 64
+    /* Let's only suppor the 64 suffix syscalls for 64-bit time_t.
+     * This simplifies the code for us as we don't need to convert
+     * between 64-bit and 32-bit times and check for overflows.
```

Comment:
Overflows?  Conversion from 32-bit to 64-bit is safe, so changing the _timespec_ to its 64 bit equivalent is okay.  Conversion back would need to happen but should be safe assuming that _timeout_ is reduced as per `select(2)` -- the values would always be smaller in magnitude.  From the man page, it's not clear how _timeout_ is modified: there is no mention of it changing but it isn't declared _const_ .

Reply:
The overflow comes because __NR_io_pgetevents_time64 will return a 64-bit time. If the platform is using a 32-bit time_t then the 64-bit value will need to be obtained from the syscall (to ensure alignment of the struct) but then converted to a 32-bit value. This is where the overflow could happen and we would need to check for it.

It's not hard to do, but I don't think it's required as eventually the 32-bit platforms will all be 64-bit time_t (before 2038).

Reply:
Even without overflows, we need to create a 64-bit version of `timespec` to make sure the struct aligns. If you feel it's worth doing we can, but I'm not sure it's worth it for a call that probably won't be called.

Reply:
Don't these calls return the number of events not a time?  The time comes back in the `timespec` and it will only be less than the value passed.

Agreed, that there would need to be a 64-bit version of `timespec`.

Reply:
Sorry, when I say return I mean values being set in `timespec`.

It seems dangerous to assume that there won't be an overflow when setting `timespec`.

Do you want me to support a 32-bit time_t and use a 64-bit version of `timespec`?

Reply:
glibc doesn't yet have `__timespec64` but it's coming.

I don't see a 64-bit `timespec` in musl, so I'm not sure how easy it will be to support this. If there isn't a `__timespec64` then this will be a compile error even for platforms with a 64-bit time_t.